In [1]:
import pandas as pd
import numpy as np
import json

In [2]:
df_real = pd.read_csv('../Uncleaned Data/Real_Scrapped.csv')
print(df_real.shape)
df_real.head()



(909, 3)


url  \
0  https://www.cnn.com/2020/03/29/politics/trump-...   
1  https://www.cnn.com/world/live-news/coronaviru...   
2  https://www.cnn.com/2020/03/29/politics/trump-...   
3  https://www.cnn.com/interactive/2020/health/co...   
4  https://www.cnn.com/2020/03/29/opinions/donald...   

                                               title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
3                  Tracking Covid-19 cases in the US   
4                         Donald Trump's special day   

                                                text  
0  Washington (CNN) President Donald Trump said S...  
1  The exterior of Elmhurst Hospital in Queens on...  
2  (CNN) President Donald Trump acknowledged Sund...  
3                                                NaN  
4  Sign up to get our new weekly column as a news...

In [3]:
df_real['label'] = 1
df_real.dropna(inplace=True)
print(df_real.shape)
df_real.drop(['url'],axis=1,inplace=True)
df_real.head()

(905, 4)


title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
4                         Donald Trump's special day   
5             How long will Trump's poll surge last?   

                                                text  label  
0  Washington (CNN) President Donald Trump said S...      1  
1  The exterior of Elmhurst Hospital in Queens on...      1  
2  (CNN) President Donald Trump acknowledged Sund...      1  
4  Sign up to get our new weekly column as a news...      1  
5  (CNN) The 2020 race, publicly at least, has la...      1

In [4]:
df_fake = pd.read_csv('../Uncleaned Data/fake.csv')

In [5]:
print(df_fake.shape)
df_fake.head()

(12999, 20)


uuid  ord_in_thread  \
0  6a175f46bcd24d39b3e962ad0f29936721db70db              0   
1  2bdc29d12605ef9cf3f09f9875040a7113be5d5b              0   
2  c70e149fdd53de5e61c29281100b9de0ed268bc3              0   
3  7cf7c15731ac2a116dd7f629bd57ea468ed70284              0   
4  0206b54719c7e241ffe0ad4315b808290dbe6c0f              0   

                 author                      published  \
0     Barracuda Brigade  2016-10-26T21:41:00.000+03:00   
1  reasoning with facts  2016-10-29T08:47:11.259+03:00   
2     Barracuda Brigade  2016-10-31T01:41:49.479+02:00   
3                Fed Up  2016-11-01T05:22:00.000+02:00   
4                Fed Up  2016-11-01T21:56:00.000+02:00   

                                               title  \
0  Muslims BUSTED: They Stole Millions In Gov’t B...   
1  Re: Why Did Attorney General Loretta Lynch Ple...   
2  BREAKING: Weiner Cooperating With FBI On Hilla...   
3  PIN DROP SPEECH BY FATHER OF DAUGHTER Kidnappe...   
4  FANTASTIC! TRUMP'S 7 POINT PLAN To Reform Heal...   

                                                text language  \
0  Print They should pay all the back all the mon...  english   
1  Why Did Attorney General Loretta Lynch Plead T...  english   
2  Red State : \nFox News Sunday reported this mo...  english   
3  Email Kayla Mueller was a prisoner and torture...  english   
4  Email HEALTHCARE REFORM TO MAKE AMERICA GREAT ...  english   

                         crawled             site_url country  domain_rank  \
0  2016-10-27T01:49:27.168+03:00  100percentfedup.com      US      25689.0   
1  2016-10-29T08:47:11.259+03:00  100percentfedup.com      US      25689.0   
2  2016-10-31T01:41:49.479+02:00  100percentfedup.com      US      25689.0   
3  2016-11-01T15:46:26.304+02:00  100percentfedup.com      US      25689.0   
4  2016-11-01T23:59:42.266+02:00  100percentfedup.com      US      25689.0   

                                        thread_title  spam_score  \
0  Muslims BUSTED: They Stole Millions In Gov’t B...       0.000   
1  Re: Why Did Attorney General Loretta Lynch Ple...       0.000   
2  BREAKING: Weiner Cooperating With FBI On Hilla...       0.000   
3  PIN DROP SPEECH BY FATHER OF DAUGHTER Kidnappe...       0.068   
4  FANTASTIC! TRUMP'S 7 POINT PLAN To Reform Heal...       0.865   

                                        main_img_url  replies_count  \
0  http://bb4sp.com/wp-content/uploads/2016/10/Fu...              0   
1  http://bb4sp.com/wp-content/uploads/2016/10/Fu...              0   
2  http://bb4sp.com/wp-content/uploads/2016/10/Fu...              0   
3  http://100percentfedup.com/wp-content/uploads/...              0   
4  http://100percentfedup.com/wp-content/uploads/...              0   

   participants_count  likes  comments  shares  type  
0                   1      0         0       0  bias  
1                   1      0         0       0  bias  
2                   1      0         0       0  bias  
3                   0      0         0       0  bias  
4                   0      0         0       0  bias

In [6]:
df_fake = df_fake[['site_url','title','text']]
df_fake.dropna(inplace=True)
df_fake.rename(columns={"site_url": "url"}, inplace=True)
chosen_idx = np.random.choice(df_fake.shape[0], replace=False, size=df_real.shape[0])
df_fake_trim = df_fake.iloc[chosen_idx]

In [7]:
df_fake_trim['label'] = 0
print(df_fake_trim.shape)
df_fake_trim.drop(['url'], axis=1, inplace= True)
df_fake_trim.head()


(905, 4)


/Users/jpyneni/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/Users/jpyneni/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4102: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


title  \
2117   The Daily Traditionalist: Speaking of Independ...   
10748  Leaked Email: Hillary Admits That Clinton Dono...   
3512   Wikileaks: Bill Clinton BOASTS of Hillary’s ‘W...   
11154  Keiser Report: Observations on US elections (E...   
11539                                        Nasty Women   

                                                    text  label  
2117   Radio Aryan October 26, 2016 \nDr Matthew Raph...      0  
10748  In a leaked email between the Democratic Candi...      0  
3512   Print \nThe bombshells about this criminal are...      0  
11154  17 mins ago 1 Views 0 Comments 0 Likes Drone f...      0  
11539    Recipient Email => \nDonald Trump grabbed a ...      0

In [8]:
df_grover = pd.read_csv('../Uncleaned Data/Neural_Fake_News.csv')
print(df_grover.shape)
df_grover.dropna(inplace=True)
print(df_grover.shape)
df_grover.drop(['Unnamed: 0'], axis=1, inplace= True)
df_grover.head()
df_grover['label'] = 2
df_grover.head()

(909, 3)
(908, 3)


title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
3                  Tracking Covid-19 cases in the US   
4                         Donald Trump's special day   

                                                text  label  
0  President Donald Trump, on his first official ...      2  
1  By Dylan Welch\nApril 1, 2015. (Bloomberg) -- ...      2  
2  Donald Trump won his home state of California ...      2  
3  Editor’s note: Readers and visitors in the New...      2  
4  The wish of many in the press was for the morn...      2

# Experiment Number 1
## Naive Bayes, Real Human News and Fake Human News
905 instances of each 

Class: Label

Real: 0

Fake: 1

In [9]:
#combine both the real and trimmed fake news into one df
combined = pd.concat([df_real,df_fake_trim],ignore_index=True)

In [10]:
print(combined.head())
combined.tail()

                                               title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
3                         Donald Trump's special day   
4             How long will Trump's poll surge last?   

                                                text  label  
0  Washington (CNN) President Donald Trump said S...      1  
1  The exterior of Elmhurst Hospital in Queens on...      1  
2  (CNN) President Donald Trump acknowledged Sund...      1  
3  Sign up to get our new weekly column as a news...      1  
4  (CNN) The 2020 race, publicly at least, has la...      1  


title  \
1805  Scientists Say Universe Is Part Of 4th Dimensi...   
1806  Saturn’s Baffling Hexagon — The Raging Storm F...   
1807           The Spiritual Alchemy of Every Day Stuff   
1808  Severe Allergies to Peanuts, Eggs and Dairy Di...   
1809  Cartoon of the Day: What lurks in the swamp Tr...   

                                                   text  label  
1805  A new scientific theory claims our universe wa...      0  
1806  NASA’s Cassini spacecraft discovered a static ...      0  
1807  The Spiritual Alchemy of Every Day Stuff Nov 2...      0  
1808  . Severe Allergies to Peanuts, Eggs and Dairy ...      0  
1809  Clinton aide’s for-profit firm illegally raise...      0

In [11]:
import nltk
from wordcloud import WordCloud
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, accuracy_score, recall_score, precision_score

In [12]:
combined_text = combined['text'].values
X_title =combined['title'].values
labels = combined['label'].values
combined.groupby('label').agg('count')

title  text
label             
0        905   905
1        905   905

In [13]:
tfidf = TfidfVectorizer(ngram_range=(1,2), max_df= 0.85, min_df= 0.01)

In [14]:
combined_text = tfidf.fit_transform(combined_text)
# combined['title'] = tfidf.fit_transform (combined['title'])

In [15]:
print(type(combined_text))
print("tf idf vectorized text shape: {}".format(combined_text.shape))
print("original text shape: {}".format(combined['text'].shape))


<class 'scipy.sparse.csr.csr_matrix'>
tf idf vectorized text shape: (1810, 10045)
original text shape: (1810,)


In [16]:
train_text, test_text, train_labels, test_labels  = train_test_split(combined_text, labels, test_size=0.2)


In [17]:
# print(train_text)
print(type(train_text))
print("size of train_text: {}".format(train_text.shape))

# print(train_labels)
print(type(train_labels))
print("size of train_labels: {}".format(train_labels.shape))


<class 'scipy.sparse.csr.csr_matrix'>
size of train_text: (1448, 10045)
<class 'numpy.ndarray'>
size of train_labels: (1448,)


In [18]:
# print(test_text)
print(type(test_text))
print("size of test_text: {}".format(test_text.shape))

# print(test_labels)
print(type(test_labels))
print("size of test_labels: {}".format(test_labels.shape))



<class 'scipy.sparse.csr.csr_matrix'>
size of test_text: (362, 10045)
<class 'numpy.ndarray'>
size of test_labels: (362,)


In [19]:
from sklearn.naive_bayes import MultinomialNB
nb = MultinomialNB()


In [20]:
nb.fit(train_text,train_labels)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [21]:
train_pred = nb.predict(train_text)

In [22]:
print('Naive Bayes In Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(train_labels, train_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(train_labels, train_pred)*100))

Naive Bayes In Sample F1 and Accuracy Scores:
F1 score 96.89%
Accuracy score 96.89%


In [23]:
combined.loc[np.where(train_labels != train_pred)[0].tolist()].groupby('label').agg('count')

title  text
label             
0         18    18
1         27    27

In [24]:
test_pred = nb.predict(test_text)
print('Naive Bayes Out of Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(test_labels, test_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(test_labels, test_pred)*100))

Naive Bayes Out of Sample F1 and Accuracy Scores:
F1 score 94.47%
Accuracy score 94.48%


In [25]:
combined.loc[np.where(test_labels != test_pred)[0].tolist()].groupby('label').agg('count')

title  text
label             
1         20    20

# Experiment 2
## Augmenting with Neural Fake News

905 Real News, 905 Fake (neural and human combined)

80% Train = 1448
20% Test = 362 (1/2 Real, 1/2 Human Fake)

Train on (1/2 real news, 1/2 fake news (1/4 human fake + 1/4 neural fake))

Human Fake = 543 (362 for train, 181 for test)

Neural Fake = 362 for train

In [49]:
print(df_grover.shape)
print(df_real.shape)
print(df_fake.shape)

(908, 3)
(905, 3)
(12273, 3)


In [50]:
trim_size = int(df_real.shape[0] * 0.6)
chosen_idx = np.random.choice(df_fake.shape[0], replace=False, size=trim_size)
df_fake_trim_2 = df_fake.iloc[chosen_idx]
df_fake_trim_2['label'] = 0
print(df_fake_trim_2.shape)
df_fake_trim_2.drop(['url'], axis=1, inplace= True)
df_fake_trim_2.head()

(543, 4)


/Users/jpyneni/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


title  \
4461   Biggest Winners and Losers of the 2016 U.S. Pr...   
12585  She Died And Came Back To Life With This Incre...   
1911                      Beyond the Anti-Trump Protests   
11361  BREAKING : TED CRUZ CALLS FOR SPECIAL PROSECUT...   
6555   Heseltine regrets not strangling mad bitch. Mo...   

                                                    text  label  
4461   . Biggest Winners and Losers of the 2016 U.S. ...      0  
12585  Share on Facebook Doctors had given Anita Moor...      0  
1911   Beyond the Anti-Trump Protests November 17, 20...      0  
11361  BREAKING : TED CRUZ CALLS FOR SPECIAL PROSECUT...      0  
6555     Midfield Diamond   Midfield Diamond   0 Nove...      0

In [51]:
trim_size = int(df_real.shape[0]*0.4)
chosen_idx = np.random.choice(df_grover.shape[0], replace=False, size=trim_size)
df_grover_trim = df_grover.iloc[chosen_idx]
df_grover_trim['label'] = 0
print(df_grover_trim.shape)
df_grover_trim.head()

(362, 3)


/Users/jpyneni/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


title  \
1    Coronavirus live updates: Global pandemic kill...   
210  FEMA Says at Least 7 People at the Disaster Ag...   
598  How Republicans' tax overhaul could make a rec...   
179  Coronavirus, France, Lockdown: Your Friday Bri...   
708   NFL veteran Donald Penn puts Encino home in play   

                                                  text  label  
1    By Dylan Welch\nApril 1, 2015. (Bloomberg) -- ...      0  
210  FEMA and other federal agencies have alerted n...      0  
598  By Bill Winters and Nancy Tartaglione\nWASHING...      0  
179  Does the world need to resort to more than one...      0  
708  The NFL never gets free, and Pro Football Hall...      0

In [52]:
print(df_real.shape)
df_real.head()

(905, 3)


title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
4                         Donald Trump's special day   
5             How long will Trump's poll surge last?   

                                                text  label  
0  Washington (CNN) President Donald Trump said S...      1  
1  The exterior of Elmhurst Hospital in Queens on...      1  
2  (CNN) President Donald Trump acknowledged Sund...      1  
4  Sign up to get our new weekly column as a news...      1  
5  (CNN) The 2020 race, publicly at least, has la...      1

In [63]:
#Create Test Set first from trimmed
trim_size = int(df_real.shape[0]*.2)
mask =  np.random.choice(df_real.shape[0], trim_size, replace=False) 
neg_mask = [i for i in range(0, df_real.shape[0]) if i not in mask]
df_real_test = df_real.iloc[mask]
df_real_train = df_real.iloc[neg_mask]
print(df_real_test.shape)
print(df_real_train.shape)

(181, 3)
(724, 3)


In [68]:
#Create Test Set first from trimmed
print(df_fake_trim_2.shape)
trim_size = int(df_fake_trim_2.shape[0]*(1/3))
mask =  np.random.choice(df_fake_trim_2.shape[0], trim_size, replace=False) 
neg_mask = [i for i in range(0, df_fake_trim_2.shape[0]) if i not in mask]
df_fake_trim_2_test = df_fake_trim_2.iloc[mask]
df_fake_trim_2_train = df_fake_trim_2.iloc[neg_mask]
print(df_fake_trim_2_test.shape)
print(df_fake_trim_2_train.shape)

(543, 3)
(181, 3)
(362, 3)


In [69]:
#combine both the real and trimmed fake news into one df
combined_train = pd.concat([df_real_train,df_fake_trim_2_train,df_grover_trim],ignore_index=True)
combined_test = pd.concat([df_real_test,df_fake_trim_2_test],ignore_index=True)

In [72]:
print(combined_train.shape)
print(combined_train.head())
combined_train.tail()

(1448, 3)
                                               title  \
0  Trump extends federal social distancing guidel...   
1  Coronavirus live updates: Global pandemic kill...   
2  Trump concedes US coronavirus death toll could...   
3                         Donald Trump's special day   
4             How long will Trump's poll surge last?   

                                                text  label  
0  Washington (CNN) President Donald Trump said S...      1  
1  The exterior of Elmhurst Hospital in Queens on...      1  
2  (CNN) President Donald Trump acknowledged Sund...      1  
3  Sign up to get our new weekly column as a news...      1  
4  (CNN) The 2020 race, publicly at least, has la...      1  


title  \
1443  ‘Sanctuary’ laws are latest immigration battle...   
1444  Tornado rips through Arkansas city, injuring 2...   
1445  Trump’s refusal to use wartime powers to direc...   
1446  NASA, Air Force conduct telework test on Frida...   
1447  Here’s who qualifies for coronavirus paid sick...   

                                                   text  label  
1443  A case may have answered a part of the questio...      0  
1444  KANSAS CITY, Mo. — A tornado rips through Arka...      0  
1445  WASHINGTON — Trump’s refusal to rely on wartim...      0  
1446  Humans can learn how to quickly respond to a d...      0  
1447  A new federal law in place this year will make...      0

In [73]:
print(combined_test.shape)
print(combined_test.head())
combined_test.tail()

(362, 3)
                                               title  \
0  Tornado rips through Arkansas city, injuring 2...   
1  Coronavirus Threatens to Overwhelm Cities’ Soc...   
2  Krzysztof Penderecki, Polish Composer With Cin...   
3  ‘America’s governors’: Andrew Cuomo and Gavin ...   
4  8 Killed After Philippine Plane Bursts Into Fl...   

                                                text  label  
0  (CNN) A tornado ripped through Jonesboro, Arka...      1  
1  The economic shutdown caused by the coronaviru...      1  
2  But then, opera had been his most troublesome ...      1  
3  New York Gov. Andrew Cuomo adopted the role of...      1  
4  MANILA — A plane used as an air ambulance by t...      1  


title  \
357  Americans Preparing for Strong Possibility of ...   
358  Hillary Wants Aggressively Interventionist For...   
359  Former NATO Chief: We Need US as ‘World’s Poli...   
360  Comment on Professor who tweeted against PC cu...   
361  “Gaslighting”: One of The Most Dangerous Forms...   

                                                  text  label  
357  By Vic Bishop With the U.S. presidential elect...      0  
358  10-27-1 6 The first Bill and Hillary Clinton c...      0  
359  Complains Obama Isn't Hawkish Enough   \nIn a ...      0  
360  Posted on October 31, 2016 by DCG | 11 Comment...      0  
361  posted by Eddie The universe can be a funny pl...      0

In [80]:
train_text = combined_train['text'].values
test_text = combined_test['text'].values
train_labels = combined_train['label'].values
test_labels = combined_test['label'].values

In [88]:
type(train_text)
print(train_text.shape)
print(test_text.shape)
fit_text = np.concatenate((train_text, test_text))
print(fit_text.shape)

(1448,)
(362,)
(1810,)


In [89]:
tfidf_2 = TfidfVectorizer(ngram_range=(1,2), max_df= 0.85, min_df= 0.01)
tfidf_2.fit(fit_text)
train_text = tfidf_2.transform(train_text)
print(type(train_text))
print("tf idf vectorized text shape: {}".format(train_text.shape))
print("original text shape: {}".format(combined_train['text'].shape))

test_text = tfidf_2.transform(test_text)
print(type(test_text))
print("tf idf vectorized text shape: {}".format(test_text.shape))
print("original text shape: {}".format(combined_test['text'].shape))

<class 'scipy.sparse.csr.csr_matrix'>
tf idf vectorized text shape: (1448, 9819)
original text shape: (1448,)
<class 'scipy.sparse.csr.csr_matrix'>
tf idf vectorized text shape: (362, 9819)
original text shape: (362,)


In [90]:
nb_2 = MultinomialNB()
nb_2.fit(train_text,train_labels)
train_pred = nb_2.predict(train_text)

In [91]:
print('Naive Bayes In Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(train_labels, train_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(train_labels, train_pred)*100))

Naive Bayes In Sample F1 and Accuracy Scores:
F1 score 91.85%
Accuracy score 91.85%


In [92]:
combined.loc[np.where(train_labels != train_pred)[0].tolist()].groupby('label').agg('count')

title  text
label             
0         63    63
1         55    55

In [93]:
test_pred = nb_2.predict(test_text)
print('Naive Bayes Out of Sample F1 and Accuracy Scores:')
print('F1 score {:.4}%'.format(f1_score(test_labels, test_pred, average='macro')*100 ))
print ('Accuracy score {:.4}%'.format(accuracy_score(test_labels, test_pred)*100))

Naive Bayes Out of Sample F1 and Accuracy Scores:
F1 score 91.68%
Accuracy score 91.71%


In [94]:
combined.loc[np.where(test_labels != test_pred)[0].tolist()].groupby('label').agg('count')

title  text
label             
1         30    30